torch/csrc/jit/ir.h

```diff
@@ -60,6 +61,67 @@ static inline bool operator==(const Use & a, const Use & b) {
 // Graph holds a list of parameters.
 struct Param;
 
+// Scope is a node of a trie that represents the tree of nested scopes.
+// Individual scopes are pushed and popped from Graph, which hold a
+// pointer to the current scope. Each Node in Graph holds a pointer
+// to the scope that was current when the node was created.
+// The trie never needs to shrink, it only grows until it is disposed
+// of when Graph is deallocated. Hence, pointers to scopes held by nodes
+// will always be valid.
+// Note that the ownership model as specified in the destructor implies 
+// that you cannot share common subtries between tries.
+struct Scope {
+private:
+  Scope* parent_;
+  Symbol name_;
+  std::unordered_set<const Scope*> children_;
+public:
+  Scope() {
+    name_ = stringToSymbol("");
+    parent_ = NULL;
+  }
+  Scope(Scope* parent, Symbol name) {
+    name_ = name;
+    parent_ = parent;
+  }
+  ~Scope() {
+    for (auto* child: children_) delete child;
+  }
+  Scope* push(Symbol name) {
+    Scope* newScope = new Scope(this, name);
+    children_.insert(newScope);
+    return newScope;
+  }
+  Scope* pop() {
+    if (parent_ == NULL) {
+      return this;
+    }
+    return parent_;
+  }
+  bool isRoot() {
+    return parent_ == NULL;
+  }
+  Scope* getRoot() {
+    Scope* current = this;
+    while (current->parent_) {
+      current = current->parent_;
+    }
+    return current;
+  }
+  Symbol name() {
+    return name_;
+  }
+  std::vector<Symbol> namesToRoot() {
+    std::vector<Symbol> out;
+    Scope* current = this;
+    while (!current->isRoot()) {
+      out.push_back(current->name_);
+      current = current->parent_;
+    }
+    return out;
+  }
```

Comment:
Yeah, I was reversing before concatenating in Graph. Now I chose to just output the concatenated string directly, since I'm not using the flat list of scope nodes to Root for anything else.